### ISSUE TITLE
RDS Modern (full) not completing

### ISSUE THREAD

**Issue by meholley** [tedsmith/NSRL-Stripper#2]
Hello,

I found your tool and have used it to convert the most recently (2021-08-25) NSRLs -> SHA-1 hash sets. THe IOS, Android, and Legacy NSRLs convert successfully, but for some reason the Modern (full) NSRL will not convert to SHA-1, it consistently halts at 3,670.467 values written. I have tried it on a couple of different computers and all attempts stop at the same values written.

I've tried it with both V1 and V1.1 of the application, same result.

Not sure if its something that I am doing or if this is legitimately an issue with NSRLStripper.

-Mike.

**Comment by coolcalmcollected22** [comment]
I have also observed this behavior with RDS 2.74...

**Comment by dwhite20899** [comment]
That line in the NSRLFile.txt file contains a file name string which includes three lower ASCII control characters: 013, 030 and 032. I think the 013 (vertical tab) is throwing the error.

```
0000360  \r  \n   "   0   4   F   E   4   5   D   5   6   2   B   1   3
0000400   6   0   A   3   5   E   8   7   A   E   3   8   4   3   D   B
0000420   0   0   5   9   C   F   4   3   3   8   C   "   ,   "   6   0
0000440   B   B   5   A   B   7   F   5   D   9   1   8   8   5   B   0
0000460   0   7   0   E   D   C   3   2   D   D   D   6   9   2   "   ,
0000500   "   B   0   E   C   4   F   F   4   "   ,   "   ! 030   V  \v
0000520       V   > 032   "   ,   1   3   7   6   7   6   8   ,   2   2
0000540   2   9   0   3   ,   "   3   6   2   "   ,   "   "  \r  \n   "
```

**Comment by tedsmith** [comment]
Hi both and thanks for the note here. 

I wasn't sure if many people (any people) actually used this to be honest because I never hear from anyone about it really except the very occasional note. So I don't\haven't really actively maintained it. I guess in some ways its good because it either means nobody is using it, or it is working fairly well for those that do. 

If I get chance I'll look into this. Your details @dwhite20899  are very helpful. Looking at it, I wonder if the issue here is not NSRL Stripper necessarily, but the formatting of the filename in the NSRLFile.txt? i.e. non valid chars?

**Comment by tedsmith** [comment]
There is definitely something amiss here. I tried downloading the minimal RDS set lastnight, first, before trying the bigger one, and tried that. NSRL Stripper stopped at 397,772, which was hash value 04FE45D562B1360A35E87AE3843DB0059CF4338C which the filename is '! V V >' (hex 21 18 56 0B 20 56 3E 1A), only about 20% into the file. 
![Untitled](https://user-images.githubusercontent.com/1560598/142429358-986c990d-620a-4869-be19-18f034e2acc3.png)

I'll look at the bigger one later, but the point is that NSRL Stripper is not expecting to see these chars. So I can probably resolve it by checking the strings, but they didn't exist in the datasets last time I compiled the exe.

**Comment by tedsmith** [comment]
OK, so, the problem is the source NSRL files now seem to contain some non-ASCII chars for some entries. Example : 0x18, 0x0B, 0x1A. 0x1A, for example, is CTRL-Z, which is the DOS EOF char. So, that explains a lot!

My program is getting "tricked" by these chars (most likely 0x1A), because it expects files of pure text type, that dont include non-text data and so it is thinking it has reached the end of the file when it hits these (as it uses an old fashioned EOF routine), so it is just stopping at the line containing said chars. 

I am told that even in the DOS era, relying on ^Z was pretty unusual. It was a CP/M thing (hence also CCP/M, CDOS, MDOS and so on) since the original CP/M filesystem had a filesize granularity of 128 bytes (i.e. it didn't store an exact length). There are still rare situations where a (text) file can end up being padded with ^Z, e.g. if it's been through an X-Modem transfer. But I don't know what the process used by NIST is. 

I'll try to make a change to correct for this. But I will ask NIST if including non-printable ASCII chars in the filename fields of NSRLFile.txt files is intentional, or not. I'm sure it is, and there will be a valid reason but I'm unsure what it would be right now. 

Meanwhile, I'll try and compile a new version tonight for you folks to try out.

**Comment by coolcalmcollected22** [comment]
Sounds good. I spoke to the NIST NSRL rep previously when I first started seeing the issue in case something different was being performed in their process. He confirmed the presence of those characters, though he referred to them as "low ASCII". He said they are investigating on their end to be sure that is the actual file name and noted that some vendors may be getting more inventive with file names up to and including including emojis. I think it is possible this could happen again in some form or fashion.

**Comment by tedsmith** [comment]
@pcstopper18 nicely done. That is good to know. Thank you. I too have pinged them a note and I am told they might follow this bug report themselves which is good of them. 

Anyway, this issue should now be resolved in the latest v1.2 release that I have just compiled and pushed out to master, and the binary is in the [releases section](https://github.com/tedsmith/NSRL-Stripper/releases/tag/v1.2). I tested it myself on the full 26Gb data set, and it finished fully, creating a 9Gb output file of SHA-1 values. It did take nearly 4 hours but I was only running in a Windows virtual machine. It shouldn't take that long on a higher spec machine. The good news is it never exceeded 5Mb of RAM. Pleasing. 

Anyway, thank you all for the engagement and helpful reports and data. Hopefully this is now resolved.
